### ISSUE TITLE
Consider structural changes for data privacy and growth

### ISSUE THREAD

**Issue by nas5w** [nas5w/semantic-validation#3]
A couple thoughts on the structure of the actual codebase:

1) Currently, the `SemanticValidator` class exposes properties like `currentProp` and `validationRules`, which the user would never need to see (they only need to be accessible to methods in the class). We should consider implementing the [Revealing Module Pattern](https://stackoverflow.com/questions/5647258/how-to-use-revealing-module-pattern-in-javascript) to create a closure and prevent any unnecessary variables from being exposed

2) We should consider designing more modularly from the beginning... [this pattern](https://stackoverflow.com/a/38124640/6525724) seems reasonable. This would allow the various methods to be grouped into their own files rather than one big main file. Perhaps the file structure could end up something like:

```
.src
+-- index.js
+-- stringValidation
|   +-- contains.js
|   +-- emailFomat.js
|   +-- stringLength.js
|   +-- ...
+-- numberValidation
|   +-- ...
+ -- etc...
```

**Comment by akuny** [comment]
Re 1, I agree that separating public and private methods is a good idea, and that the revealing module pattern is a good way of doing so while we're working with a single file for the `SemanticValidator` class. 

Re 2, would each individual file (for example, stringLength.js), follow a convention of exposing strictly one public method and >=0 private methods?

**Comment by nas5w** [comment]
I put together a quick branch [here](https://github.com/nas5w/semantic-validation/blob/modular-files/src/index.js) that shows both the revealing module pattern and individual file pattern seemingly working well together. I also changed the `addRule` method name to `selectProp` and then created an actual `addRule` method that does what it sounds like.

The one problem I see with it is it'd be better to make the `shouldContain` function pure (i.e., not reference `this.addRule`) since that reference creates a dependency.

**Comment by nas5w** [comment]
This has been implemented in the `master` branch. That being said, I think the best option for the time being is to make the "private" members of the `SemanticValidator` module public. Here's what I have encountered that makes me think this is currently the best option:

- There is currently no such thing as a "private" variable in JS classes ([it's being planned](https://github.com/tc39/proposal-class-fields#private-fields) for some future ECMAScript standard.)
- We are attempting to work around this fact by using an IIFE to create a closure, making certain variables/methods (i.e., `currentProp`, `validationRules`, `addRule`) only accessible to the `SemanticValidator` class.
- Doing this appeared to work at first glance, but it the IIFE only happens once, meaning there is only one instance of these variables in memory. This means that, if _multiple_ validator objects are instantiated, they will be referencing the same "private" variables/methods in memory! In any real OO implementation, these private properties would be created for each instance.
- This is not a problem for a pure, stateless function like `addRule`, but a big problem for the stateful `currentProp` and `validationRules` variables that really are supposed to belong to each instance of the class.
- It is not at all unlikely to see a scenario in the wild where multiple validator objects will exist at once--I bet it would be very common in SPAs.
- There is a [potential workaround](https://stackoverflow.com/a/28165599/6525724). In this example, you create a variable in the constructor and also re-define each method to reference the newly-created variable. This would probably work but seems incredibly inefficient, especially as the number of validation methods grows.

For all these reasons, I think it'd be best to define the "private" variables/functions as prototypal properties/methods on the `SemanticValidator` class. We can just not document them in the public API docs and perhaps even abstract them to a "core" file with comments not to use them.

**Edit:** I should mention the way I realized all of this. My test suite was failing in a very confusing way. Since a new `SemanticValidator` object is instantiated for each test, by the final test it turned out there were about 10 rules in the `validationRules` array. This is because each new instance of `SemanticValidator` was using, and adding to, the `validationRules` array used by the previous instances. Yikes!
